### PR TITLE
WIP/ENH: Make S3File iterable, readline

### PR DESCRIPTION
**Pull request by TomAugspurger**
Adds a `readline` method to S3File, which is used to make them
iterable via py2 style `.next` and py3 style `__next__` methods.

### CODE REVIEW

File: s3fs/tests/test_s3fs.py [hunk 1]
```diff
@@ -79,6 +80,75 @@ def test_simple(s3):
...
+        f.write(data)
+    with s3.open(a, 'rb') as f:
+        result = r.readline()
+        assert result == data
+
+def test_next(s3):
+    expected = csv_files['2014-01-01.csv'].split(b'\n')[0] + b'\n'
+    with s3.open(test_bucket_name + '/2014-01-01.csv') as f:
+        result = next(f)
+        assert result == expected
```
Comment (mrocklin) [inline s3fs/tests/test_s3fs.py]: Might want to try `__iter__` as well.  Something like the following:

    assert list(f) == csv_files['2014-01-01.csv'].split(b'\n')

File: s3fs/tests/test_s3fs.py [hunk 2]
```diff
@@ -79,6 +80,75 @@ def test_simple(s3):
...
+
+def test_readline_empty(s3):
+    data = b''
+    with s3.open(a, 'wb') as f:
+        f.write(data)
+    with s3.open(a, 'rb') as f:
+        out = f.readline()
+        assert out == b''
+
```
Comment (mrocklin) [inline s3fs/tests/test_s3fs.py]: Do these tests stress blocksize?

File: s3fs/tests/test_s3fs.py (hunk 1 above)
Comment (TomAugspurger) [inline s3fs/tests/test_s3fs.py]: I think I pushed a second commit after you made this note testing something like that https://github.com/dask/s3fs/pull/18/commits/a27d2916feba0743b8c75127e6a4049d2bf36b9b

File: s3fs/tests/test_s3fs.py (hunk 2 above)
Comment (TomAugspurger) [inline s3fs/tests/test_s3fs.py]: They do not, and it looks like I don't handle the case where the line is longer than blocksize correctly. Fixing now.

**Comment by TomAugspurger** [comment]
https://github.com/dask/s3fs/pull/18/commits/d5ab2dfdedc40022e8014a7db8774771f65864d3 should take care of the blocksize issue.

**Comment by TomAugspurger** [comment]
The last commit added some file-like properties `readable`, `writeable`, and `seekable`. These were enough to pass most of the pandas test-suite (which doesn't have too many s3 tests admittedly). I was briefly looking into making `S3File` a proper subclass of one of the AbstractBaseClasses in `io`, but this seems to be enough for now.

There are a couple more pandas issues I'm working around.

- We had some tests for `s3n` and `s3a` [URLs](https://wiki.apache.org/hadoop/AmazonS3). I'm not sure whether pandas was properly handling those in the past, we probably just treated them like regular s3 urls.
- Maybe something pandas dealt with [here](https://github.com/pydata/pandas/pull/10604). Basically a public bucket that has a public objects sitting next to private objects. When we go to list the objects in the directory, we get a permissions exception.

Will followup tonight hopefully. I can keep throwing changes in the PR if you want, or followup with another.

**Comment by mrocklin** [comment]
Whatever makes you happier.  I'm really quite glad that you in particular
are handling this.

**Comment by martindurant** [comment]
Please compare with my version in PR #19.
The readline function there is somewhat simpler, and I believe copes with cases where a seek precedes a readline.
Interesting to see the requirements for readable etc.

**Comment by TomAugspurger** [comment]
The need for readable, seekable, etc. are probably peculiar to pandas. We've got the C parser, and a fallback parser written in Python. For the Python parser, I needed to wrap the byte stream from here in a `TextIOWrapper`, which required those methods to be present.

Close this one in favor of #19?

**Comment by martindurant** [comment]
If you're happy with #19, then yes.

**Pull request closed without merging**